### PR TITLE
Add analyzer option to limit the number of TUs imported

### PR DESCRIPTION
**Pull request by gamesh411**

### CODE REVIEW

File: lib/StaticAnalyzer/Core/AnalyzerOptions.cpp [hunk 1]
```diff
@@ -54,9 +54,9 @@ AnalyzerOptions::UserModeKind AnalyzerOptions::getUserMode() {
     StringRef ModeStr =
         Config.insert(std::make_pair("mode", "deep")).first->second;
     UserMode = llvm::StringSwitch<UserModeKind>(ModeStr)
-      .Case("shallow", UMK_Shallow)
-      .Case("deep", UMK_Deep)
-      .Default(UMK_NotSet);
+                   .Case("shallow", UMK_Shallow)
```
Comment (martong) [inline lib/StaticAnalyzer/Core/AnalyzerOptions.cpp]: Is this change related?

**Review by martong — COMMENTED** [review]
You have so many unrelated changes which seems like the result of an automated formatting. Could you please avoid them?
You may use `git clang-format` which formats only the changed lines.

**Review by martong — APPROVED** [review]
Looks good!

**Comment by gamesh411** [comment]
I have run into a dependency issue. If I store the TU threshold as an AnalyzerOption, then I have to link with libStaticAnalyzerCore (see CMakeLists.txt change). But this introduces a cyclic dependency. I see a couple of solutions to this problem:

a) put the option elsewhere, for example as a general frontend-option
b) make the check elsewhere, for example inside some analyzer-specific manager class
c) try to cut the dependency by introducing intermediary libs

What are your thoughts?

**Comment by martong** [comment]
> I have run into a dependency issue. If I store the TU threshold as an AnalyzerOption, then I have to link with libStaticAnalyzerCore (see CMakeLists.txt change). But this introduces a cyclic dependency. I see a couple of solutions to this problem:
> 
> a) put the option elsewhere, for example as a general frontend-option
> b) make the check elsewhere, for example inside some analyzer-specific manager class
> c) try to cut the dependency by introducing intermediary libs
> 
> What are your thoughts?

I prefer option a). Because CrossTU is ought to be independent from the analyzer, that is more generic.

**Review by martong — APPROVED** [review]
Looks good!

**Comment by Szelethus** [comment]
Shouldn't this be an analyzer config option instead? I am aware of several CTU options that are just that, why is this a frontend option?

**Comment by martong** [comment]
> Shouldn't this be an analyzer config option instead? I am aware of several CTU options that are just that, why is this a frontend option?

Yes this should be an analyzer option, but the first attempt somehow created a lib  dependency failure.
So, Endre is working on the second attempt where he will extend the parameters of `CrossTranslationUnitContext::getCrossTUDefinition` like we do with `DisplayCTUProgress`.

**Review by martong — CHANGES_REQUESTED** [review]
Please extend the parameters of CrossTranslationUnitContext::getCrossTUDefinition like we do with DisplayCTUProgress.

**Comment by Szelethus** [comment]
> > Shouldn't this be an analyzer config option instead? I am aware of several CTU options that are just that, why is this a frontend option?
> 
> Yes this should be an analyzer option, but the first attempt somehow created a lib dependency failure.

The reason probably was that my analyzer config changes are not live on clang7 -- the solution would be to leave the option's getter method in-class, I believe.

**Comment by dkrupp** [comment]
deadline for this is 31st of March!

**Comment by gamesh411** [comment]
Reorganized parameter handling, added unit test and some documentation.

**Comment by Szelethus** [comment]
Sorry for missing out on the discussion. I think adding this as an analyzer-config flag WOULD be preferred on clang 8.0.0, but since we can't list and verify these options in 7.0.0., let's not shoot ourselves in the foot with them.

LGTM regarding option placement, I can't say much about the rest of the patch.

File: include/clang/StaticAnalyzer/Core/AnalyzerOptions.h [hunk 2]
```diff
@@ -186,6 +186,10 @@ class AnalyzerOptions : public RefCountedBase<AnalyzerOptions> {
   // Cap the stack depth at 4 calls (5 stack frames, base + 4 calls).
   unsigned InlineMaxStackDepth = 5;
 
+  /// The maximal amount of translation units that is considered for import
+  /// when inlining functions during CTU analysis.
+  unsigned CTUImportThreshold{100u};
```
Comment (martong) [inline include/clang/StaticAnalyzer/Core/AnalyzerOptions.h:191]: I am not sure that aggregate initialization will be compiled without problems on GCC 4.8. Clang master still supports that compiler with C++11, right?
Perhaps ` = 100u;` would be more appropriate.

**Review by martong — APPROVED** [review]
LG! Nice work!

**Pull request merged**